custom_transpose2.py

```diff
@@ -74,43 +75,87 @@ def __call__(self, res_arg, lin_arg):
     call_jaxpr = core.ClosedJaxpr(pe.convert_constvars_jaxpr(jaxpr), ())
 
     assert self._fun_transpose is not None, 'todo error'
+    flat_rule = _flatten_rule(
+      lu.wrap_init(self._fun_transpose), res_tree, lin_tree, out_tree())
+
     out_flat = custom_transpose_p.bind(*consts, *args_flat,
                                        call=call_jaxpr,
-                                       rule=self._fun_transpose,
+                                       rule=flat_rule,
                                        num_consts=len(consts),
                                        res_tree=res_tree,
                                        lin_tree=lin_tree,
                                        out_tree=out_tree())
     return tree_unflatten(out_tree(), out_flat)
-  
+
+@lu.transformation
+def _flatten_rule(res_tree, lin_tree, out_tree, *args):
+  assert len(args) == res_tree.num_leaves + out_tree.num_leaves
+  res_flat, cts_out_flat = util.split_list(args, [res_tree.num_leaves])
+  res = tree_unflatten(res_tree, res_flat)
+  cts_out = tree_unflatten(out_tree, cts_out_flat)
+  cts_in = yield (res, cts_out), {}
+  # TODO(danfm): Handle Nones in cts_in
+  cts_in_flat, cts_in_tree = tree_flatten(cts_in)
```

Comment:
Would "handling Nones" amount to this?
```suggestion
  cts_in_flat, cts_in_tree = tree_flatten(cts_in, is_leaf=lambda x: x is None)
```

Reply:
I think yes!